dialect/sql/schema/postgres.go

```diff
@@ -332,15 +347,241 @@ func (d *Postgres) addColumn(c *Column) *sql.ColumnBuilder {
+		} else {
+			actualEnums[typ] = &NativeEnum{
+				Name:   typ,
+				Values: []string{label},
+			}
+		}
+	}
+
+	changes := &enumChanges{}
+
+	for enumName, expectedEnum := range expectedEnums {
+		// Modify
+		if actualEnum, ok := actualEnums[enumName]; ok {
+			actualEnum.Diff = actualEnum.Values.diff(expectedEnum.Values)
+
+			if !actualEnum.Diff.equal() {
+				changes.modify = append(changes.modify, actualEnum)
+			}
+		} else {
+			// Add new enum
+			changes.add = append(changes.add, expectedEnum)
+		}
+	}
+
+	// Find enums to delete
+	for enumName, v := range actualEnums {
+		if _, ok := actualEnums[enumName]; !ok {
+			changes.drop = append(changes.drop, v)
+		}
+	}
+
+	return changes, nil
+}
+
+func (d *Postgres) createEnums(ctx context.Context, tx dialect.Tx, nativeEnums []*NativeEnum) error {
+	for _, ne := range nativeEnums {
+		var result sql.Result
+		err := tx.Exec(ctx, fmt.Sprintf("CREATE TYPE %s as ENUM ('%s')", ne.Name, strings.Join(ne.Values, "','")), []interface{}{}, result)
+
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+const enumColumnsQuery = `
+select col.table_name,
+       col.column_name,
+       col.udt_name
+from information_schema.columns col
+         join information_schema.tables tab on tab.table_schema = col.table_schema
+    and tab.table_name = col.table_name
+    and tab.table_type = 'BASE TABLE'
+         join pg_type typ on col.udt_name = typ.typname
+where col.table_schema = current_schema()
+  and typ.typtype = 'e'
+order by col.table_schema,
+         col.table_name,
+         col.ordinal_position;
+`
+
+type columnsUsingEnum struct {
+	tableName string
+	colName   string
+}
+
+func (d *Postgres) alterEnums(ctx context.Context, tx dialect.Tx, nativeEnums []*NativeEnum) error {
+	if len(nativeEnums) == 0 {
+		return nil
+	}
+
+	rows := &sql.Rows{}
+	err := tx.Query(ctx, enumColumnsQuery, []interface{}{}, rows)
+
+	if err != nil {
+		return err
+	}
+
+	defer rows.Close()
+
+	enumsInUse := make(map[string][]columnsUsingEnum)
+
+	for rows.Next() {
+		var (
+			tableName string
+			colName   string
+			enumName  string
+		)
+		if err := rows.Scan(&tableName, &colName, &enumName); err != nil {
+			return fmt.Errorf("scanning index description %v", err)
+		}
+
+		if _, ok := enumsInUse[enumName]; ok {
+			enumsInUse[enumName] = append(enumsInUse[enumName], columnsUsingEnum{
+				tableName: tableName,
+				colName:   colName,
+			})
+		} else {
+			enumsInUse[enumName] = []columnsUsingEnum{{
+				tableName: tableName,
+				colName:   colName,
+			}}
+		}
+	}
+
+	typeTempName := func(name string) string {
+		return fmt.Sprintf(" %s_entmigrate", name)
+	}
+
+	// Step 1. Rename types to be modified to <type>_entmigrate
+	for typeName := range enumsInUse {
+		var result sql.Result
+		err := tx.Exec(ctx, fmt.Sprintf("ALTER TYPE %s RENAME TO %s", typeName, typeTempName(typeName)), []interface{}{}, result)
+
+		if err != nil {
+			return err
+		}
+	}
+
+	// Step 2. Create new Enums
```

Comment:
Can we use "new enum" strategy only as a fallback?

In situation where only new values are added to already existing enum, we can use just `ALTER TYPE ADD VALUE` .

Reply:
alter type add value cannot run in a transaction. So if the migration failed we will leave the database in a 'dirty' state

Reply:
Seems like it was changed recently?

> If ALTER TYPE ... ADD VALUE (the form that adds a new value to an enum type) is executed inside a transaction block, the new value cannot be used until after the transaction has been committed.

https://www.postgresql.org/docs/current/sql-altertype.html

Probably we can leave it as is and introduce opt-in optimization for that.

Reply:
@ernado thats great information. Looks like its been added since 12.

Reply:
I will be happy to add that in a followup pr. this one is already quite big